### PR TITLE
create base class for SR2ML models

### LINKED ISSUE #26: Common Base Class for SR2ML Models
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
More and More models are added to SR2ML, they all share some common functionalities, it is better to create a  Common Base Class for them to improve the maintainability of the code. 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #26 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual.
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Regression tests have to complete successfully.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test.
- [x] 6. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.

### CODE REVIEW

File: src/Bases/ModelBase.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      This method is utility function that tells if the reliability model is able to
+      treat dynamic data on its own or not
+      @ In, None
+      @ Out, isDynamic, bool, True if the reliability model is able to treat dynamic data, False otherwise
+    """
+    return self._dynamicHandling
+
+  def setVariable(self, value):
+    """
+      Set value if a float/int/list is provided in the node text, othersise treat the provided value as RAVEN variable
```
Comment (mandd) [inline src/Bases/ModelBase.py]: otherwise

File: src/Bases/ModelBase.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    else:
+      # should be floats; InputData assures the entries are the same type already
+      if not utils.isAFloatOrInt(value[0]):
+        raise IOError('Multiple non-number entries are found, but require either a single variable name or multiple float entries: {}'.format(value))
+      ret = np.asarray(value)
+    return ret
+
+  def loadVariables(self, need, inputDict):
+    """
+      Load the values of variables that is generated by RAVEN
```
Comment (mandd) [inline src/Bases/ModelBase.py]: are

File: src/Bases/ModelBase.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    """
+      Set the parameters from a given dictionary.
+      @ In, paramDict, dict, settings
+      @ Out, None
+    """
+    for key, val in paramDict.items():
+      if key in self.__dict__.keys():
+        setattr(self, key, val)
+      else:
+        print('WARNING: Variable "{}" is not defined in class "{}"!'.format(key, self.name))
```
Comment (mandd) [inline src/Bases/ModelBase.py]: should we raise a KeyError?

File: src/Bases/ModelBase.py (hunk 1 above)
Comment (wangcj05) [reply src/Bases/ModelBase.py]: updated. Thanks

File: src/Bases/ModelBase.py (hunk 2 above)
Comment (wangcj05) [reply src/Bases/ModelBase.py]: changed

File: src/Bases/ModelBase.py (hunk 3 above)
Comment (wangcj05) [reply src/Bases/ModelBase.py]: I change it to KeyError now. For our current applications, it is more safe to do so. In the future, if our models also want to acquire the meta data information, and this data was not a self variable in the model, we can remove the check. I put the check there is mainly for the developer to remind them if they forget to add the on-the-fly updated variable to self._variableDict.

**Review by mandd — APPROVED** [review]

**Pull request merged**